### PR TITLE
feat(glm5next): reduce speculative state memory with atomic LMCache recovery

### PR DESCRIPTION
**Pull request by voipmonitor**
## What changes

Supported GLM-5.3 MTP/DFlash serving automatically uses native B12X KDA state
recovery. Instead of retaining a full recurrent matrix per proposed token, it
keeps one FP32 checkpoint plus compact update records and commits only accepted
tokens. There is no additional state quantization.

**LMCache remains supported.** For atomic request-boundary checkpoints, the
runner commits accepted recurrent state and compacts convolution history before
capturing the external checkpoint. Export then uses zero speculative offset.
Full-state decoding retains its existing capture-before-compaction order. The
LMCache connector already transfers complete physical pages with layout-aware
identity, so its disk format does not need to change.

With the Spark checkpoint on two GPUs, three MTP draft tokens and a 3072-token
batch/chunk budget, fixed per-request reserve
falls from **856.58 to 668.81 MiB per rank (−21.9%)**. This is a planner reserve,
not a claim that every idle chat occupies those bytes. Wider shared pages offset
the saving for one maximum-length request: fitted context remains 897,024 tokens.

## Compatibility

- Automatic on eligible SM12x, runner V2, PP1, BF16 activation, FP32 recurrent
  state, head dimension 128, MTP/DFlash depth 1–7 configurations.
- `--no-use-replayssm` retains full speculative states. Other models stay opt-in.
- Generic aligned KV connectors and P/D transfer remain unsupported; only an
  atomic request-boundary connector can export recovery state.
- Existing multimodal generation and native prefix caching remain available.
  The atomic LMCache adapter already excludes image-bearing requests; external
  multimodal checkpoint reuse is not introduced by this PR.
- The metadata backend retains vLLM's `TRITON` name; GLM speculative verify and
  state commit execute B12X CuTe kernels.

## Validation

Qualified: **28 vLLM GPU/config/runner tests**, including the full-state and
recovery capture orders, graph replay, automatic selection and explicit opt-out.
Four CPU cases additionally check logits-only checkpoint reuse. The companion
B12X suite passes 131 tests with 29 unrelated skips, including binding follow-up
[B12X #409](https://github.com/local-inference-lab/b12x/pull/409). B12X #408 has
been merged independently by Luke; this PR remains against `dev/karmic-kraken`.

The two-GPU Spark/MTP configuration and four-GPU NVFP4 configurations with either
three MTP or seven DFlash2 draft tokens pass RAM restores, restart/disk restores,
changed prompt tails, and generated-response continuation. External probes
require zero GPU-cache hits. All three pass vision generation; the two-GPU
configuration also passes mixed prefill/decode. Matched full-state controls are
complete: TP2 C1 output changes −0.2%, C4 +0.03%, prefill +1.5%; TP4 MTP C1
−3.1%, C8 −1.8%, prefill −0.3%; DFlash2 has similar throughput.
This saves memory but has a measured TP4 MTP throughput cost; it is not
qualified as zero-regression. A suspected repetitive TP2 recovery sample is
retained and identified in the linked report, not used as clean quality evidence.
These are bounded functional checks, not a general model-accuracy evaluation.

[Source identities, measurements, limits and test receipts](https://github.com/local-inference-lab/rtx6kpro/blob/master/benchmarks/glm53-kda-recovery-lmcache.md).

```bash
OMP_NUM_THREADS=1 /opt/venv/bin/python -m pytest --noconftest \
  tests/models/glm5next/test_kda_recoverssm.py \
  tests/models/glm5next/test_kda_recurrent.py \
  tests/v1/worker/test_gpu_model_runner_v2.py \
  tests/v1/worker/test_mamba_utils.py \
  -k 'recoverssm or recovery or boundary_checkpoint or recovered_accepted or recurrent' -q
```

Changed-file lint/format hooks pass. Full mypy reports the pre-existing
`GPUModelRunner.jit_warmup_registry` attribute error, reproduced on the untouched
base; that hook is not claimed to pass.

## Attribution and overlap

Builds on Jiangyun Zhu's Kimi-K3 recovery contract (vLLM `70afdedc108`) and reuses
its metadata/convolution integration. Developed with OpenAI Codex assistance.

Checked LIL and upstream open PRs: LIL #809 is prompt-tail retention, #798 is a
broader serving-contract port, and upstream #56082 tunes Kimi recovery. None
implements this native B12X GLM recovery path with atomic LMCache export. Those
changes are not superseded.

## Summary by CodeRabbit

* **New Features**
  * Added automatic B12X state recovery for eligible GLM-5.3 speculative decoding configurations.
  * LMCache checkpoints now preserve accepted recurrent state after recovery and compaction.
  * Qualified TP2 Spark MTP3 configurations reduce fixed per-request memory reservations by approximately 188 MiB per GPU, without reducing maximum context length.
  * Multimodal generation and native GPU prefix caching remain unchanged.
  * Use `--no-use-replayssm` to select the full-state diagnostic path.

* **Bug Fixes**
  * Improved state handling and checkpoint ordering for speculative decoding recovery.

### CODE REVIEW

**Comment by voipmonitor** [comment]
Depends on local-inference-lab/b12x#408 and its binding-contract follow-up #409. Two-GPU Spark MTP3 and four-GPU MTP3/DFlash2 LMCache RAM/restart-disk checks are complete, including matched full-state performance controls. The PR body and linked evidence retain the measured MTP throughput cost. Changes and dependency-checked changelog fragments are pushed to integration/karmic-kraken-beta; the published-image qualification remains pending. Please review automatic eligibility and checkpoint capture ordering. @coderabbitai review

File: vllm/v1/attention/backends/gdn_attn.py [hunk 1]
```diff
@@ -198,6 +198,9 @@ def __init__(
         else:
             self.num_spec = 0
         self.use_spec_decode: bool = self.num_spec > 0
+        self.state_index_columns = (
+            1 if vllm_config.cache_config.use_kda_recoverssm else self.num_spec + 1
+        )
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/v1/attention/backends/gdn_attn.py]: Confirmed with a minimal CPU reproducer: a three-token speculative batch produced one-column indices for both GDNAttentionMetadataBuilder and KimiK3ROCmKDAMetadataBuilder. AMD fused_recurrent_kda indexes every token column, so that shape violates its contract. Fixed in 66bcca4e462: narrowing now requires an explicit builder capability, enabled only for the NVIDIA Kimi recovery builder and GLM recovery builder. Full-state builders retain num_spec + 1 columns even when the global flag is true. Both reproducer cases fail before the fix and pass afterward; the targeted GLM/Kimi/runner suite passes 31 tests (75 deselected). AMD GPU execution was not tested. GLM native recovery computation and selected metadata shape are unchanged.

File: vllm/model_executor/layers/mamba/gdn/kimi_gdn_linear_attn.py [hunk 2]
```diff
@@ -1317,6 +1339,7 @@ def _run_b12x_kda_decode_post_conv(
         query_start_loc: torch.Tensor,
         num_accepted_tokens: torch.Tensor | None,
         num_requests: int,
+        apply_output_norm: bool = True,
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/model_executor/layers/mamba/gdn/kimi_gdn_linear_attn.py:1301]: Added apply_output_norm to the Args section in 66bcca4e462, including its default. No redundant caller comment was added. Changed-file lint, formatting and mypy hooks pass.

**Comment by voipmonitor** [comment]
Consolidated #824 into this PR, preserving both histories and its immutable release fragment. Recovery records are forwarded through the alias-safe binding; the conflicting gate/output case uses disjoint output until the gate has been consumed. This is byte-identical to the qualified beta implementation, with compatible test fixtures. The composed KK stack passes 81 configuration/offload/recovery tests and 112 QSA/GDN/warmup/alias tests (nine declared skips). Merge #798 first: its workspace and warmup interfaces are prerequisites. The standalone branch type check reports those missing prerequisite interfaces; the composed runtime retains them. #824 will be closed as superseded to avoid conflicting duplicate merges.
